here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -160,6 +163,81 @@ class WhereClauseBuilder(private val request: ReadFeatures) {
         }
     }
 
+    private fun whereTags() {
+        val tagQuery = request.query.tags
+        if(tagQuery != null){
+            if (where.isNotEmpty()) {
+                where.append(" AND (")
+            } else {
+                where.append(" (")
+            }
+            whereNestedTags(tagQuery)
+            where.append(")")
+        }
+    }
+
+    private tailrec fun whereNestedTags(tagQuery: ITagQuery) {
```

Comment:
`tailrec` works on compilation level, and it will try to replace recursion with `while` loop during compilation - that's all.

Reply:
There are couple of things to consider here
- `tailrec` does not need function to marked as `inline`
- `tailrec` as keyword is a compiler hint, nothing more - by itself it does nothing
- for the `tailrec` hint to work, the function must satisfy some other rules for the compile-time optimization take place 

The last point is where I screwed up ;) `tailrec` won't be applied here because function **must end with self-call** for optimization to take place. So here, the `tailrec` keyword won't do anythign. It won't help, it won't throw error.
i removed it, it will use simple recursion explicitly.

source: https://kotlinlang.org/docs/functions.html#tail-recursive-functions